An Introduction to CI/CD Tools (#2041)
section-engineering-education/engineering-education: articles/an-introduction-to-cicd-tools/index.md

- [Buddy as a CI/CD Tool](#buddy-as-a-ci/cd-tool)
- [Conclusion](#conclusion)
- [Further Reading and References](#further-reading-and-references)

### Technique of CI/CD
Continuous Integration (CI) is a programming practice requiring developers to incorporate code changes into a shared repository. Changes committed to the repository are verified by an automated construction, ensuring that bugs are spotted early before deployment.

Understanding the CI process in-depth, whenever a change is made to the source repository in a version control system, an automated process builds the framework. It runs a series of tests to ensure the recent code update integrates with the existing code in the repository. If it doesn’t, then a conflict error occurs.

When a conflict error occurs between the newly integrated code and the existing code, it is difficult to track the source of the conflict and fix it. However, with the aid of Continuous Delivery, developers in large organizations can securely and rapidly organize changes of all sorts, e.g., features upgrade, interface changes, bug fixes, and tests in production environments or to deploy sustainably to the end-user.

> The technique behind the integration of CI/CD in the software development process is that in a large organization with teams of thousands of developers making changes regularly, we can still ensure that the software doesn’t break and changes are being synchronized seamlessly, therefore ensuring effective communication between developers and operations.

Reviewer: @corpsgeek This sentence is too long making it hard to read. Please try splitting it into two sentences

![ci/cd technique](/engineering-education/an-introduction-to-cicd-tools/technique.png)

### The Importance of CI/CD
- CI reduces the number of processes that are repeated across all project operations, including code compilation, database integration, testing, inspection, deployment, and feedback.
- The core importance of Continuous Integration is to ensure that software being developed is deployable to end-users, but with better quality and lower risks of failure. 
- CI gives you the ability to spot trends and make informed decisions, as well as the courage to make new changes. When there is no actual or recent data to support decisions, projects suffer, and everyone makes their best guesses.
- Successful implementation of CI practices can boost confidence in software development, such that the team will know that test checks are passed, project code and design follow development standards, and the result is a functioning testable product with each build.

### Jenkins as a CI/CD Tool
Jenkins is an open-source automation server written in Java. It is a server-based continuous integration application that requires a web server to operate on.

> 
> In this article, we'll provide a structural approach to using Jenkins in automating the entire deployment process and using a docker container to deploy our code to the server. The whole development cycle will be based on the DevOps model with the help of DevOps tools as well. 

Reviewer: ```suggestion 
In this article, we'll provide a structural approach to using Jenkins to automate the entire deployment process and use a docker container to deploy our code to the server. The whole development cycle will be based on the DevOps model with DevOps tools' help. 
```

![Jenkins cycle view](/engineering-education/an-introduction-to-cicd-tools/jenkinspipe.png)

The following steps illustrate the generic approach of implementing Jenkins:

> This article assumes that you have a prior understanding of implementing CI/CD tools and thereby reinforces that knowledge to adequately provide a step-by-step approach to using one.

Reviewer: ```suggestion
This article assumes that you have a prior understanding of implementing CI/CD tools, thereby reinforcing that knowledge to provide a step-by-step approach to using one adequately.
```

To use Jenkins with your software, you can either build a git repository or use an existing one for your CI/CD. The development team is in charge of committing the code to the Dev-Branch.

> 
> When changes are adequately committed to the Dev-Branch, Jenkins can download the source code from Github and have it mapped to a job that is configured for a particular role. Once a job is configured, you must ensure that continuous integration and continuous development are completed for the job/task.

Reviewer: ```suggestion
When changes are adequately committed to the Dev-Branch, Jenkins can download the source code from Github and mapped to a configured job for a particular role. Once a job is configured, you must ensure that continuous integration and continuous development are completed for the job/task.
```

- Exceptionally adaptable
- Package uploading with automated merging and custom commands
- Setup is easy, and you can create as many times as you want.

### Implementing CircleCI
To leverage the features of CircleCI, the version control system to be used should either be GitHub or Bitbucket. When you add a new project to CircleCI, you would be required to set some permission parameters, and that's to your preference as a developer. 

However, when you add a new project, you have to add the following cloud settings:
- A deploy key is required for testing the project from Bitbucket Cloud or GitHub.
- A service hook is required by GitHub or Bitbucket cloud to alert CircleCI.

By default, CircleCI creates a service hook. As a result, builds are activated for all repository service hook, with PUSH being the most common cause of triggering a build.

### Bamboo as a CI/CD tool
Bamboo is a continuous integration server that creates a continuous delivery pipeline by automating the management of software application releases. It also covers the building of tests, assigning versions, tagging releases, deploying and activating new versions on the production server.

> Developers can use Bamboo integration to automate construction and test processes for any software program. It also provides a platform for segregating builds with different targets and specifications. Another benefit is the ability to automatically deploy the program into the server for release, which saves a lot of time when compared to manual testing. 

Reviewer: ```suggestion
Developers can use Bamboo integration to automate construction and test processes for any software program. It also provides a platform for segregating builds with different targets and specifications. Another benefit is the ability to automatically deploy the program into the server for release, saving a lot of time compared to manual testing. 
```

### Implementing Bamboo
Bamboo's workflow is fairly straightforward in the process of coordinating builds and test suites. The configuration of its workflow order of actions is usually divided into different segments and sections that make up the workflow's building blocks. 

The Bamboo workflow building blocks are: 
- Plans
- Jobs
- Tasks

#### The Plan Block 
By default, plans have only one stage. It can, however, be used to divide jobs into different stages and then execute them as needed. For faster execution and app tests, plans use the same repository to run a series of one or more stages sequentially. Plans ensure that you have everything in one place, even though you have separate stages to separate jobs and tasks within them.

#### The Jobs Block
Jobs are a collection of tasks that are executed in a specific order on the same agent. They give you control over the order in which the tasks for your build must be completed. Jobs gather task requirements to map out the skills needed in the Bamboo CI Server's comprehension.

> #### The Task Block
> Tasks are the smallest discrete working units that carry out any system's command, such as parsing test results, executing maven goals, running scripts, and executing source code checkouts are examples of general tasks within a job on the working Bamboo directory. These tasks run in order. 

Reviewer: ```suggestion
Tasks are the smallest discrete working units that carry out any system's command, such as parsing test results, executing maven goals, running scripts, and executing source code checkouts. These are examples of general tasks within a job of the working Bamboo directory. These tasks run in order.
```

> ### Buddy as a CI/CD tool
> Buddy is a continuous integration and delivery (CI/CD) tool that builds, tests, and deploys websites and applications using code from GitHub, Bitbucket, and GitLab. It uses Docker containers with pre-installed languages and frameworks, as well as DevOps, tracking, and notification activities, to build on.

Reviewer: ```suggestion
Buddy is a continuous integration and delivery (CI/CD) tool that builds, tests, and deploys websites and applications using code from GitHub, Bitbucket, and GitLab. It uses Docker containers with pre-installed languages and frameworks and DevOps, tracking, and notification activities to build on.
```
